### ISSUE TITLE
[BUG] High CPU usage by instance manager

### ISSUE THREAD

**Issue by HonakerM** [longhorn/longhorn#3636]
## Describe the bug
This is a continuation of [this bug report](https://github.com/longhorn/longhorn/issues/1533#issuecomment-1038601469). I'm seeing 100% cpu usage by a `tgtd` process. 

## To Reproduce

1. Install longhorn via [helm](https://longhorn.io/docs/1.2.3/deploy/install/install-with-helm/)
2. Provision volumes with the `longhorn` storage class
3. Volumes will be successfully bound and created; however, a pod tried to mount the volume it will fail 
4. The node's instance manager will then expand to use 100% of the CPU

## Expected behavior
 Have a low CPU usage from pod instance-manager-e-xxxx (lower than consuming 4 CPU)

## Log or Support bundle

Here is a list of all of my longhorn pods and their resource usage (note instance-manager-e-167eff13 ):
```
NAME                                        CPU(cores)   MEMORY(bytes)
csi-attacher-6454556647-6xdl4               0m           15Mi
csi-attacher-6454556647-z52zh               0m           9Mi
csi-provisioner-869bdc4b79-l6t6w            0m           9Mi
csi-provisioner-869bdc4b79-nwrnm            0m           10Mi
csi-provisioner-869bdc4b79-znn4g            2m           14Mi
csi-resizer-6d8cf5f99f-gsx4p                0m           8Mi
csi-resizer-6d8cf5f99f-qm5h7                0m           15Mi
csi-resizer-6d8cf5f99f-zngh7                0m           12Mi
csi-snapshotter-588457fcdf-6pnbx            2m           11Mi
csi-snapshotter-588457fcdf-kd9mt            0m           13Mi
csi-snapshotter-588457fcdf-nr5zf            0m           9Mi
engine-image-ei-fa2dfbf0-dh46h              5m           9Mi
engine-image-ei-fa2dfbf0-f2c54              4m           12Mi
engine-image-ei-fa2dfbf0-w5db5              1m           6Mi
instance-manager-e-167eff13                 3433m        71Mi
instance-manager-e-50d209a9                 3m           17Mi
instance-manager-e-54542bca                 9m           22Mi
instance-manager-r-26157459                 3m           44Mi
instance-manager-r-b8b0cead                 89m          37Mi
instance-manager-r-e0818bc8                 32m          58Mi
longhorn-csi-plugin-cvm9s                   0m           15Mi
longhorn-csi-plugin-rbcvn                   2m           19Mi
longhorn-csi-plugin-t5cx8                   0m           22Mi
longhorn-driver-deployer-7dddcdd5bb-zxdzl   0m           8Mi
longhorn-manager-6mmxp                      40m          69Mi
longhorn-manager-7lcrs                      8m           87Mi
longhorn-manager-qmdhj                      59m          75Mi
longhorn-ui-7648d6cd69-zqkql                0m           2Mi
```

When viewing the logs of the `instance` I'm seeing alot of the following two snippets:
```
[longhorn-instance-manager] time="2022-02-17T17:05:37Z" level=info msg="wait for process pvc-d410a8a2-c885-4e95-a122-1d84c1ed5427-e-f398b362 to shutdown"
[longhorn-instance-manager] time="2022-02-17T16:42:51Z" level=debug msg="Process Manager: wait for process pvc-21fddd09-3d84-4cfb-8bc4-94300d22a2c5-e-d2e573b4 to shutdown before unregistering process"
[longhorn-instance-manager] time="2022-02-17T16:42:51Z" level=debug msg="Process Manager: wait for process pvc-fde0f0df-9391-430f-af5c-6023139337c0-e-71e498e1 to shutdown before unregistering process"
[longhorn-instance-manager] time="2022-02-17T16:42:51Z" level=debug msg="Process Manager: wait for process pvc-21fddd09-3d84-4cfb-8bc4-94300d22a2c5-e-d2e573b4 to shutdown before unregistering process"
[longhorn-instance-manager] time="2022-02-17T16:42:52Z" level=debug msg="Process Manager: wait for process pvc-fde0f0df-9391-430f-af5c-6023139337c0-e-71e498e1 to shutdown before unregistering process"
[longhorn-instance-manager] time="2022-02-17T16:42:52Z" level=debug msg="Process Manager: wait for process pvc-21fddd09-3d84-4cfb-8bc4-94300d22a2c5-e-d2e573b4 to shutdown before unregistering process"
[longhorn-instance-manager] time="2022-02-17T16:42:52Z" level=debug msg="Process Manager: wait for process pvc-21fddd09-3d84-4cfb-8bc4-94300d22a2c5-e-d2e573b4 to shutdown before unregistering process"
```
and I'm also seeing alot of the following `tgtd` errors
```
tgtd: abort_cmd(1325) found 10000068 6
tgtd: iscsi_tcp_show(552) Transport endpoint is not connected
tgtd: conn_close_admin(253) close 1f4 0
tgtd: tgt_event_modify(271) Cannot find event 515
tgtd: iscsi_event_modify(575) tgt_event_modify failed
```

The main command taking up all of the resources is the following `tgtd` process:
```
tgtd -f
 ```
 
## Environment

 - Longhorn version: 1.2.3
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: K8s v1.23.1
   - Number of management node in the cluster: 1 (marked as schedulable)
   - Number of worker node in the cluster: 2
 - Node config
   - OS type and version: Ubuntu 20.04
   - CPU per node: 4CPU on management, 2CPU on workers
   - Memory per node: 2gb
   - Disk type(e.g. SSD/NVMe): HDD
   - Network bandwidth between the nodes: all are VMs on the same host
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Proxmox
 - Number of Longhorn volumes in the cluster: 4

## Additional Information
This issue can be temporary resolved by restarting the node; however, the issue will always return on one of the nodes.

**Comment by vavallee** [comment]
Having this issue also, following.

**Comment by derekbit** [comment]
@HonakerM 
Could you provide more details including CPU model and underlying storage (1 single HDD or RAID)?

**Comment by HonakerM** [comment]
@derekbit I've seen this happen on both of my clusters. One is 3vms that all share a pair of raid 1 drives. The other cluster utilizes network attached isci storage (I'm not sure the underlying storage).

I was trying to focus on the first cluster since it's the one I have most control over.

**Comment by derekbit** [comment]
@HonakerM 
Can you provide the support bundle and kubelet log for further analysis?

**Comment by HonakerM** [comment]
When I try to get the bundle through the longhorn ui it hangs forever with no error (I've waited hours). Is there a way to generate it from the cli?

**Comment by derekbit** [comment]
@HonakerM 
You can refer to  https://longhorn.io/kb/troubleshooting-create-support-bundle-with-curl/.

**Comment by HonakerM** [comment]
@derekbit  Please find the requested logs attached.
[longhorn-support-bundle_b1ef9d0b-d1e4-4049-8024-dcfeb0a0dfb3_2022-02-22T03-25-40Z.zip](https://github.com/longhorn/longhorn/files/8113487/longhorn-support-bundle_b1ef9d0b-d1e4-4049-8024-dcfeb0a0dfb3_2022-02-22T03-25-40Z.zip)

**Comment by derekbit** [comment]
Many R/W timeout events in longhorn-engine log.
Probably caused by the disk starvation.

```
2022-02-18T00:12:12.395164887Z Feb 18 00:11:59.116368: ->172.16.13.104:10001 P[  0kB]      203us
2022-02-18T00:12:12.395169768Z Feb 18 00:12:01.112607: ->172.16.13.104:10001 P[  0kB]      221us
2022-02-18T00:12:12.395174363Z Feb 18 00:12:03.112639: ->172.16.13.104:10001 P[  0kB]      247us
2022-02-18T00:12:12.395178957Z Feb 18 00:12:05.112573: ->172.16.13.104:10001 P[  0kB]      221us
2022-02-18T00:12:12.395183469Z Feb 18 00:12:07.112660: ->172.16.13.104:10001 P[  0kB]     1035us
2022-02-18T00:12:12.395188181Z Feb 18 00:12:09.112617: ->172.16.13.104:10001 P[  0kB]      270us
2022-02-18T00:12:12.395192814Z Feb 18 00:12:00.966357: ->172.16.231.208:10046 R[  4kB]  8151924us failed
2022-02-18T00:12:12.395216546Z Feb 18 00:05:01.122244: ->172.16.231.208:10046 P[  0kB] 427996051us failed
2022-02-18T00:12:12.395222925Z Feb 18 00:05:00.471599: ->172.16.200.243:10046 P[  0kB] pending
2022-02-18T00:12:15.693718293Z timeout_handler: Timeout request 508 due to disconnection
2022-02-18T00:12:16.165023856Z tgtd: bs_longhorn_request(105) fail to read at 2181267456 for 4096
2022-02-18T00:12:16.165041721Z tgtd: bs_longhorn_request(149) io error 0x56231cbb4520 28 -14 4096 2181267456, Success
2022-02-18T00:12:16.165438673Z timeout_handler: Timeout request 509 due to disconnection
2022-02-18T00:12:16.165458642Z tgtd: bs_longhorn_request(91) fail to write at 0 for 4096
2022-02-18T00:12:16.165465018Z tgtd: bs_longhorn_request(149) io error 0x56231cbb4520 2a -14 4096 0, Success
2022-02-18T00:12:17.760361839Z [pvc-9ec931c0-4d85-48e3-8428-a519d816160c-e-bd4344bf] time="2022-02-18T00:12:17Z" level=error msg="R/W Timeout. No response received in 8s"
2022-02-18T00:12:17.760385019Z Feb 18 00:12:11.112605: ->172.16.13.104:10001 P[  0kB]      272us
2022-02-18T00:12:17.760392572Z Feb 18 00:12:13.112583: ->172.16.13.104:10001 P[  0kB]      319us
2022-02-18T00:12:17.760397560Z Feb 18 00:12:15.112602: ->172.16.13.104:10001 P[  0kB]      213us
2022-02-18T00:12:17.760402667Z Feb 18 00:12:17.112571: ->172.16.13.104:10001 P[  0kB]      209us
2022-02-18T00:12:17.760407333Z Feb 18 00:05:00.471599: ->172.16.200.243:10046 P[  0kB] 437288325us failed
2022-02-18T00:12:17.760413357Z Feb 18 00:12:09.120416: ->172.16.200.243:10046 R[  4kB]  8639522us failed
2022-02-18T00:12:17.760417941Z time="2022-02-18T00:12:17Z" level=error msg="Replicator.ReadAt:2 r/w timeout"
2022-02-18T00:12:17.760422788Z time="2022-02-18T00:12:17Z" level=error msg="Error reading from wire: read tcp 172.16.13.98:43568->172.16.200.243:10046: use of closed network connection"
```

**Comment by HonakerM** [comment]
@derekbit could these timeouts cause `tgtd` to suddenly use 100% of the cpu? If so is there a way this be mitigated without changing out the underlying disks?

**Comment by innobead** [comment]
cc @keithalucas

**Comment by keithalucas** [comment]
@HonakerM Can you look at `dmesg` and see if any processes on the nodes with `tgtd` using 100% of the CPU are killed because of an Out of Memory issue?  Looking at the support bundle logs it looks like `tgtd` may be reading and writing to a partially closed socket.   The longhorn-engine controller is the other side of the tgtd socket.  I was able to induce the 100% CPU usage behavior by `kill -9`ing the longhorn-engine controller process which is similar to the kernel killing it by running out of memory.

**Comment by keithalucas** [comment]
By attaching gdb, it looks like one thread in tgtd is stuck calling [read_full](https://github.com/rancher/liblonghorn/blob/9eaf16c13dccc7f3bdb111d4e21662ae753bdccf/src/longhorn_rpc_protocol.c#L8) from liblonghorn.   I'm seeing SIGPIPE too which makes sense.

**Comment by derekbit** [comment]
> By attaching gdb, it looks like one thread in tgtd is stuck calling [read_full](https://github.com/rancher/liblonghorn/blob/9eaf16c13dccc7f3bdb111d4e21662ae753bdccf/src/longhorn_rpc_protocol.c#L8) from liblonghorn. I'm seeing SIGPIPE too which makes sense.

Do you mean the iowait caused the CPU usage high?

**Comment by HonakerM** [comment]
@keithalucas I'm not seeing anything in `dmesg` probably because I had to restart to solve the issue.  Please let me know if there are any other ways I can help debug.

**Comment by keithalucas** [comment]
> > By attaching gdb, it looks like one thread in tgtd is stuck calling [read_full](https://github.com/rancher/liblonghorn/blob/9eaf16c13dccc7f3bdb111d4e21662ae753bdccf/src/longhorn_rpc_protocol.c#L8) from liblonghorn. I'm seeing SIGPIPE too which makes sense.
> 
> Do you mean the iowait caused the CPU usage high?

No, I recompiled liblonghorn with debugging symbols and believe I figured out exactly what's going on.

1. The longhorn engine has a signal handler that is executed when the engine (the `longhorn controiller` command) is signaled to exit gracefully, i.e. SIGTERM or SIGINT.  https://github.com/longhorn/longhorn-engine/blob/051ebde20b02cdeb11ff49588f5bf8e7d527c7eb/app/cmd/controller.go#L105 established the handler and https://github.com/longhorn/longhorn-engine/blob/051ebde20b02cdeb11ff49588f5bf8e7d527c7eb/pkg/controller/control.go#L787 will execute the `tgtadm` command to remove the volume from `tgtd`.   If the frontend isn't shutdown gracefully, `tgt` will continue to try to `read` and `write` from the Unix domain socket between `tgtd` and the longhorn engine.

       
2. https://github.com/rancher/liblonghorn/blob/9eaf16c13dccc7f3bdb111d4e21662ae753bdccf/src/longhorn_rpc_protocol.c#L13 alway returns 0 when the socket is closed on the other end.     liblonghorn creates a thread that reads from the Unix domain socket continuously:  https://github.com/rancher/liblonghorn/blob/9eaf16c13dccc7f3bdb111d4e21662ae753bdccf/src/longhorn_rpc_client.c#L310  It relies on the `read` syscall either blocking or returning a number > 0.   When the socket is closed on the other end, `read` returns 0 and the loop in `read_full` is an infinite loop utilizing 100% of the CPU because `readed` will never be greater than `len`
3. `tgtd` is also using liblonghorn to write when is it in this 100% CPU state.  The `write_full` function in liblonghorn is calling the `write` system call which is failing by returning -1 and setting `errno` to EPIPE.  `tgtd` logs this problem with a message like this:
```
send_msg: fail to write magic version
tgtd: bs_longhorn_request(105) fail to read at 5368643584 for 4096
tgtd: bs_longhorn_request(149) io error 0x5555555dc0f0 28 -22 4096 5368643584, Broken pipe
```

That message was in one of the instance-manager-e logs and I could recreate it.

We need to handle `read` returning 0 in `read_full` in a more appropriate manner.   We need to detect the disconnected socket and stop reading and writing to it.

**Comment by vavallee** [comment]
great news. looking forward to 1.3.0

**Comment by keithalucas** [comment]
To duplicate without kubernetes:
1. Setup longhorn by following https://github.com/longhorn/longhorn/wiki/How-to-run-longhorn-engine-without-containers
2. Find the `./longhorn-engine controller` process and `kill -9` its PID.
3. Look at the CPU utilization of `tgt` in `top`.

To duplicate with kubernetes:
1. Create a longhorn volume and attach it to a node.
2. Go into the `instance-manager-e-` POD on the node that the volume is attached to.
3. Run `ps ax` and find the PID of the `/engine-binaries/longhornio-longhorn-engine-master-head/longhorn controller` process.
4. `kill -9` the process.
5. Run `top` and look at the CPU utilization of `tgt`.  Note: longhorn can relaunch the engine process on this node and make that volume available on this node again.  If that happens, the CPU utilization stop being 0%.

**Comment by HonakerM** [comment]
@keithalucas  thank you so much for your work on this. 

In regards to the following: 
> longhorn can relaunch the engine process on this node and make that volume available on this node again. If that happens, the CPU utilization stop being 100%.

Killing the longhorn instance-manager pod does not stop the `tgtd` process that is using `100%`. Is it safe to manually kill the `tgtd` process when it happens?

**Comment by keithalucas** [comment]
> Killing the longhorn instance-manager pod does not stop the `tgtd` process that is using `100%`. Is it safe to manually kill the `tgtd` process when it happens?

@HonakerM 
`tgtd` doesn't restart if you kill it manually.  It actually ends up as a Zombie process.  I don't understand how `tgtd` could remain running when you kill the pod.  Maybe the pod is restarting and it immediately gets into the `100%` state at start up?

**Comment by HonakerM** [comment]
@keithalucas let me try to duplicate the issue on my end. Thank you for your help

**Comment by khushboo-rancher** [comment]
@keithalucas Do we have the backported issues created for this or I can help creating those?

**Comment by keithalucas** [comment]
> @keithalucas Do we have the backported issues created for this or I can help creating those?

I didn't make backported github issues.

**Comment by tabnul** [comment]
I can confirm 1.2.4 RC1 fixed CPU usage issues of longhorn manager. But my scenario was different:
- backups of volumes to S3: 100% cpu usage, most of the times the backup succeed and situation recovers
- in some cases, CPU usage stays at 100% afterwards, basically pods not working anymore until manual intervention.

Can it be that impact of backups on CPU is reduced as well?

**Comment by meldafrawi** [comment]
Validation: **PASSED**

Longhorn version: `master-head`

Steps to test:

- Create a longhorn volume and attach it to a node.
- Go into the instance-manager-e- POD on the node that the volume is attached to.
- Run ps ax and find the PID of the /engine-binaries/longhornio-longhorn-engine-master-head/longhorn controller process.
- kill -9 the process.
- Run top and look at the CPU utilization of tgt. 

Note: longhorn can relaunch the engine process on this node and make that volume available on this node again.
 If that happens, the CPU utilization stop being 100%.

**Comment by wbarnard81** [comment]
I am using longhorn 1.3.1 and I am still experiencing this issue. It does not look like there is a fix yet, except for restarting the worker node? 

How can I help to resolve this issues, as I am experiencing it in production...

**Comment by innobead** [comment]
> I am using longhorn 1.3.1 and I am still experiencing this issue. It does not look like there is a fix yet, except for restarting the worker node?
> 
> How can I help to resolve this issues, as I am experiencing it in production...

@wbarnard81 Could you help create another issue w/ your case description? Thanks.

**Comment by Smartich0ke** [comment]
I'm also having this issue in longhorn 1.7.0. Restarting the worker node sometimes works. Very frustrating.

**Comment by derekbit** [comment]
> I'm also having this issue in longhorn 1.7.0. Restarting the worker node sometimes works. Very frustrating.

@Smartich0ke Can you provide more information? Which process? How to trigger the high CPU issue?

**Comment by Smartich0ke** [comment]
@derekbit I ran `top` in the affected instance manager pod. Here is a picture of the output:
![image](https://github.com/user-attachments/assets/f842199a-6700-40a7-9e0a-48d0ddc2f447)

There is nothing really needed to trigger the high CPU usage. I just boot the node and once the container starts, it starts using a lot of CPU and the fans of the node rev up.
Here are some pictures of the CPU metric from prometheus:
![image](https://github.com/user-attachments/assets/e6b2e2e9-d566-4749-9e9a-6dcbc4bcc3aa)
![image](https://github.com/user-attachments/assets/86c38691-38f7-43c6-816a-3e34e44fb71e)

This just happened spontaneously one day, I didn't make any changes to how longhorn was configured.

I have tried several reboots of the node now, and they don't seem to resolve the issue. I was also running on 1.6.3 previously, so I upgraded to 1.7.0 to make sure it hadn't been fixed. The CPU usage seems like it is definitely less on 1.7.0 on the affected node but still abnormally high.

I also noticed the RAM usage on the instance managers is quite high. about 2-3GB per pod. But this has always been like that so I suspect it's a separate issue.

The node is using a 7th gen i7.
